runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/StreamingEngineClient.java

```diff
@@ -0,0 +1,444 @@
+   */
+  public static StreamingEngineClient create(
+      JobHeader jobHeader,
+      GetWorkBudget totalGetWorkBudget,
+      GrpcWindmillStreamFactory streamingEngineStreamFactory,
+      WorkItemProcessor processWorkItem,
+      WindmillStubFactory windmillGrpcStubFactory,
+      GetWorkBudgetDistributor getWorkBudgetDistributor,
+      GrpcDispatcherClient dispatcherClient) {
+    StreamingEngineClient streamingEngineClient =
+        new StreamingEngineClient(
+            jobHeader,
+            totalGetWorkBudget,
+            new AtomicReference<>(StreamingEngineConnectionState.EMPTY),
+            streamingEngineStreamFactory,
+            processWorkItem,
+            windmillGrpcStubFactory,
+            getWorkBudgetDistributor,
+            dispatcherClient,
+            new Random().nextLong());
+    streamingEngineClient.startGetWorkerMetadataStream();
+    streamingEngineClient.startWorkerMetadataConsumer();
+    return streamingEngineClient;
+  }
+
+  @VisibleForTesting
+  static StreamingEngineClient forTesting(
+      JobHeader jobHeader,
+      GetWorkBudget totalGetWorkBudget,
+      AtomicReference<StreamingEngineConnectionState> connections,
+      GrpcWindmillStreamFactory streamFactory,
+      WorkItemProcessor processWorkItem,
+      WindmillStubFactory stubFactory,
+      GetWorkBudgetDistributor getWorkBudgetDistributor,
+      GrpcDispatcherClient dispatcherClient,
+      long clientId) {
+    StreamingEngineClient streamingEngineClient =
+        new StreamingEngineClient(
+            jobHeader,
+            totalGetWorkBudget,
+            connections,
+            streamFactory,
+            processWorkItem,
+            stubFactory,
+            getWorkBudgetDistributor,
+            dispatcherClient,
+            clientId);
+    streamingEngineClient.startGetWorkerMetadataStream();
+    streamingEngineClient.startWorkerMetadataConsumer();
+    return streamingEngineClient;
+  }
+
+  @SuppressWarnings("FutureReturnValueIgnored")
+  private void startWorkerMetadataConsumer() {
+    newWorkerMetadataConsumer.submit(
+        () -> {
+          while (true) {
+            Optional.ofNullable(newWindmillEndpoints.poll())
+                .ifPresent(this::consumeWindmillWorkerEndpoints);
+          }
+        });
+  }
+
+  /**
+   * Waits for the initial worker metadata to be consumed. Does nothing if this has already been
+   * called.
+   *
+   * @throws IllegalArgumentException if trying to start before {@link #connections} are set with
+   *     {@link GetWorkerMetadataStream}.
+   */
+  public void waitForFirstWorkerMetadata() {
+    // Do nothing if we have already initialized the initial streams.
+    if (started.compareAndSet(false, true)) {
+      return;
+    }
+
+    // This will block if we have not received the first response from GetWorkerMetadata.
+    try {
+      getWorkerMetadataReady.await();
+    } catch (InterruptedException e) {
+      throw new StreamingEngineClientException(
```

Comment:
Thread.currentThread().interrupt();

https://www.baeldung.com/java-interrupted-exception

Reply:
noted
removed await so we don't need this.